pkg/security/ebpf/c/dentry_resolver.h

```diff
@@ -0,0 +1,333 @@
+    struct dentry *d_parent;
+    struct inode *d_inode = NULL;
+    int segment_len = 0;
+
+    if (key.ino == 0 || key.mount_id == 0) {
+        return DENTRY_INVALID;
+    }
+
+#pragma unroll
+    for (int i = 0; i < DR_MAX_ITERATION_DEPTH; i++)
+    {
+        d_parent = NULL;
+        bpf_probe_read(&d_parent, sizeof(d_parent), &dentry->d_parent);
+
+        key = next_key;
+        if (dentry != d_parent) {
+            write_dentry_inode(d_parent, &d_inode);
+            write_inode_ino(d_inode, &next_key.ino);
+        }
+
+        // discard filename and its parent only in order to limit the number of lookup
+        if (input->discarder_type && i < 2) {
+            if (is_discarded_by_inode(input->discarder_type, key.mount_id, key.ino, i == 0)) {
+                return DENTRY_DISCARDED;
+            }
+        }
+
+        bpf_probe_read(&qstr, sizeof(qstr), &dentry->d_name);
+        segment_len = bpf_probe_read_str(&map_value.name, sizeof(map_value.name), (void *)qstr.name);
+        if (segment_len > 0) {
+            map_value.len = (u16) segment_len;
+        } else {
+            map_value.len = 0;
+        }
+
+        if (map_value.name[0] == '/' || map_value.name[0] == 0) {
+            map_value.name[0] = '/';
+            next_key.ino = 0;
+            next_key.mount_id = 0;
+        }
+
+        map_value.parent = next_key;
+
+        bpf_map_update_elem(&pathnames, &key, &map_value, BPF_ANY);
+
+        dentry = d_parent;
+        if (next_key.ino == 0) {
+            input->dentry = d_parent;
+            input->key = next_key;
+            return i + 1;
+        }
+    }
+
+    if (input->iteration == DR_MAX_TAIL_CALL) {
+        map_value.name[0] = 0;
+        map_value.parent.mount_id = 0;
+        map_value.parent.ino = 0;
+        bpf_map_update_elem(&pathnames, &next_key, &map_value, BPF_ANY);
+    }
+
+    // prepare for the next iteration
+    input->dentry = d_parent;
+    input->key = next_key;
+    return DR_MAX_ITERATION_DEPTH;
+}
+
+SEC("kprobe/dentry_resolver_kern")
+int kprobe__dentry_resolver_kern(struct pt_regs *ctx) {
+    struct syscall_cache_t *syscall = peek_syscall(EVENT_ANY);
+    if (!syscall)
+        return 0;
+
+    syscall->resolver.iteration++;
+    syscall->resolver.ret = resolve_dentry_tail_call(&syscall->resolver);
+
+    if (syscall->resolver.ret > 0) {
+        if (syscall->resolver.iteration < DR_MAX_TAIL_CALL && syscall->resolver.key.ino != 0) {
+            bpf_tail_call(ctx, &dentry_resolver_kprobe_progs, DR_KPROBE);
+        }
+
+        syscall->resolver.ret += DR_MAX_ITERATION_DEPTH * (syscall->resolver.iteration - 1);
+    }
+
+    if (syscall->resolver.callback >= 0) {
+        bpf_tail_call(ctx, &dentry_resolver_kprobe_callbacks, syscall->resolver.callback);
+    }
+    return 0;
+}
+
+SEC("tracepoint/dentry_resolver_kern")
+int tracepoint__dentry_resolver_kern(void *ctx) {
+    struct syscall_cache_t *syscall = peek_syscall(EVENT_ANY);
```

Comment:
it seems to be the exact same code as the kprobe one, could we make a common function called by both ?

Reply:
It will have to be a macro unfortunately: from the few tests I've made you can't have a `bpf_tail_call` on a program map that contains programs from a different type than the calling programs (even behind a if that will never be called, I guess the dead code elimination part of the verifier is done after the map coherence check ? :idk: ).

TL;DR that's why I duplicated the code, but a macro should do the trick ! Working on it ...

Reply:
👍🏻